web/frontend/architecture.md

```diff
@@ -0,0 +1,77 @@
+
+- `/src/layout` - **React components** used to render common layout elements, such as header, footer, page content, etc. Similar to the components located in `/src/components`, they should focus on rendering UI rather than application logic or connecting to providers. However, they are designed such that there should only ever be one instance on each page.
+
+- `/src/pages` - **React components** that correspond to actual routes (URLs). These are responsible for assembling the UI components for a page, and hooking them up with the necessary providers such as Redux or GraphQL. Queries should be co-located with page components, since pages are explicitly dependent on them.
+
+- `/src/store` - Redux or other application state management code is in here! For more info on whether to use Redux or GraphQL, [read here](#to-redux-or-graphql).
+
+  - `/src/store/store.ts` - Configures and exports the Redux store
+  - Directories in here correspond to reducers, or "slices" of the store. Each directory contains its own types, actions, and reducer. These should all be well-tested.
+
+- `/src/stories` - Storybook stories that aren't specific to a single component live here.
+
+- `/src/styles` - Sometimes we also have to write CSS.
+
+- `/src/types` - Type definitions that are shared throughout the application.
+
+### Application Bootstrap
+
+#### (or, how does all of this come together?)
+
+The primary entry point is `src/index.tsx`. This file is responsible for:
+
+- (if using GraphQL) Initializing ApolloClient with our configuration and any client-side mocks.
+- Including the CSS entry point (`src/index.scss`) so it doesn't feel left out, and so our webpages look pretty.
+- Wrapping the main `<App />` component inside of top-level providers, such as `<ApolloProvider>` (GraphQL), `<Provider>` (Redux), and `<PersistGate>` (redux-persist).
+- Finally, rendering the resulting `<App />` component into the DOM (!!)
+
+The next most important file is `src/App.tsx`. This is the top-level component that is rendered into the DOM. It is responsible for 2 important things:
+
+- Imposing a global `<Layout />` onto _all pages_.
+- Mapping whatever the current route is to the matching page component. If there is no matching page component, it will render the `NotFound` page component instead.
+
+## To Redux or GraphQL?
+
+A lot of confusion can stem from where to store data and how to access it in React applications. Here's a quick guide:
+
+- GraphQL is for data you are getting from the API. If you are using GraphQL, we recommend using [ApolloClient](https://www.apollographql.com/docs/react/). ApolloClient is very smart at caching, so when we get API data we don't need to store it anywhere else! Storing API data in Redux would be redundant.
+- Redux is for local application state. That means if there's any UI state that the entire application needs to be aware of, but that _doesn't_ need to be saved to a database for later, we should put it in Redux!
```

Comment:
Not sure if this is a separate discussion, but React `Context` can handle application state rather than implementing Redux in many situations.

Reply:
Definitely interested in the discussion aspect of this to learn from. I've seen some interesting benchmarks between Context, useReducer, and Redux, and a whole lot of back and forth on when to use which.

Reply:
+1 For someone who doesn't know much React this reads like you **need** to use GraphQL or Redux. I know isn't the choice we want to suggest and interestingly Truss has already had projects that didn't use either (Redox had  a large React frontend where stage was managed entirely with React.Context).

Instead of calling this "To Redux or GraphQL" the section could be titled **Managing Application State** and breifly outline what are different types of application state. Then in presenting different options we just keep these sections here and merge as is.  In future PRs we can add React.Context and ReactQuery or anything else we've seen Truss use.

Reply:
good catch all! This was from a doc I had written for SABER (which was using Redux & GraphQL) and just copied pasted. I updated to bring this section some more clarity